File: SddIA/scripts/qa/verify-tools-index.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""Cortafuegos CI: sincronización index.md <-> definiciones {name}.md (Core y local)."""

from __future__ import annotations

import os
import re
import sys
from pathlib import Path

SCRIPT = Path(__file__).resolve()
REPO_ROOT = SCRIPT.parents[3] if (SCRIPT.parents[2] / "tools").is_dir() else SCRIPT.parents[2]

CORE_DIR = REPO_ROOT / "SddIA" / "tools"
CORE_INDEX = CORE_DIR / "index.md"
LOCAL_DIR = REPO_ROOT / ".SddIA" / "tools"
LOCAL_INDEX = LOCAL_DIR / "index.md"
LEGACY_LOCAL_DIRNAME = "Tools"

EXCLUDE_STEMS = frozenset({"index", "tools-contract", "README"})
EXCLUDE_SUFFIXES = ("-contract.md",)


def _is_tool_definition(path: Path) -> bool:
    if path.suffix.lower() != ".md":
        return False
    stem = path.stem
    if stem in EXCLUDE_STEMS:
        return False
    if any(path.name.endswith(s) for s in EXCLUDE_SUFFIXES):
        return False
    return True


def _scan_definitions(tools_dir: Path) -> set[str]:
    if not tools_dir.is_dir():
        return set()
    return {p.name for p in tools_dir.iterdir() if p.is_file() and _is_tool_definition(p)}


def _parse_index_filenames(index_path: Path) -> set[str]:
    if not index_path.is_file():
        return set()
    text = index_path.read_text(encoding="utf-8")
    in_catalog = False
    indexed: set[str] = set()
    for line in text.splitlines():
        if "## Catálogo" in line or "## Catalogo" in line:
            in_catalog = True
            continue
        if in_catalog and line.startswith("## ") and "Catálogo" not in line and "Catalogo" not in line:
            break
        if not in_catalog or not line.strip().startswith("|"):
            continue
        if re.match(r"^\|\s*[-:]+", line):
            continue
        if "Archivo fuente" in line or line.strip() == "| name |":
            continue
        cells = [c.strip() for c in line.strip().strip("|").split("|")]
        if not cells:
            continue
        first = cells[0].strip()
        m = re.search(r"`([^`]+)`", first)
        token = (m.group(1) if m else first).strip()
        if not token or token.lower() in ("archivo fuente", "name"):
            continue
        if not token.endswith(".md"):
            token = f"{token}.md"
        indexed.add(token)
    return indexed


def _audit_legacy_tools_dirs(repo_root: Path) -> list[str]:
    errors: list[str] = []
    for dot in repo_root.rglob(".SddIA"):
        if not dot.is_dir():
            continue
        try:
            names = {n for n in os.listdir(dot) if (dot / n).is_dir()}
        except OSError:
            continue
        if LEGACY_LOCAL_DIRNAME in names:
            legacy = dot / LEGACY_LOCAL_DIRNAME
            errors.append(
                f"Violación de Simetría Fractal: {legacy} — exigido `{dot / 'tools'}` en minúscula estricta"
            )
    return errors


def _audit_scope(label: str, tools_dir: Path, index_path: Path) -> list[str]:
    errors: list[str] = []
    if not tools_dir.is_dir() and not index_path.is_file():
        return errors
    if index_path.is_file() and not tools_dir.is_dir():
        errors.append(f"{label}: index exists but directory missing: {tools_dir}")
        return errors
    if tools_dir.is_dir() and not index_path.is_file():
        errors.append(f"{label}: definitions present but index missing: {index_path}")
        return errors

    on_disk = _scan_definitions(tools_dir)
    in_index = _parse_index_filenames(index_path)

    for name in sorted(on_disk - in_index):
        errors.append(f"{label}: orphan file (not in index): {tools_dir / name}")
    for name in sorted(in_index - on_disk):
        errors.append(f"{label}: orphan index row (no file): {name}")
    return errors


def main() -> int:
    errors: list[str] = []
    errors.extend(_audit_legacy_tools_dirs(REPO_ROOT))
    errors.extend(_audit_scope("core", CORE_DIR, CORE_INDEX))
    if LOCAL_INDEX.is_file() or LOCAL_DIR.is_dir():
        errors.extend(_audit_scope("local", LOCAL_DIR, LOCAL_INDEX))

    if errors:
        print("verify-tools-index: FAILED", file=sys.stderr)
        for e in errors:
            print(e, file=sys.stderr)
        return 1
    print("verify-tools-index: OK")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: SddIA/tools/sddia-qa/src/verify_tools_index.rs
use regex::Regex;
use std::fs;
use std::path::Path;

const EXCLUDE_STEMS: &[&str] = &["index", "tools-contract", "README"];
const EXCLUDE_SUFFIXES: &[&str] = &["-contract.md"];
const LEGACY_LOCAL_DIRNAME: &str = "Tools";

fn is_tool_definition(path: &Path) -> bool {
    if path.extension().and_then(|e| e.to_str()) != Some("md") {
        return false;
    }
    let stem = path.file_stem().and_then(|s| s.to_str()).unwrap_or("");
    if EXCLUDE_STEMS.contains(&stem) {
        return false;
    }
    let name = path.file_name().and_then(|s| s.to_str()).unwrap_or("");
    if EXCLUDE_SUFFIXES.iter().any(|s| name.ends_with(s)) {
        return false;
    }
    true
}

fn scan_definitions(tools_dir: &Path) -> std::collections::HashSet<String> {
    let mut out = std::collections::HashSet::new();
    let Ok(rd) = fs::read_dir(tools_dir) else {
        return out;
    };
    for entry in rd.flatten() {
        let path = entry.path();
        if path.is_file() && is_tool_definition(&path) {
            if let Some(name) = path.file_name().and_then(|s| s.to_str()) {
                out.insert(name.to_string());
            }
        }
    }
    out
}

fn parse_index_filenames(index_path: &Path) -> std::collections::HashSet<String> {
    let Ok(text) = fs::read_to_string(index_path) else {
        return std::collections::HashSet::new();
    };
    let re_backtick = Regex::new(r"`([^`]+)`").expect("regex");
    let mut in_catalog = false;
    let mut indexed = std::collections::HashSet::new();
    for line in text.lines() {
        if line.contains("## Catálogo") || line.contains("## Catalogo") {
            in_catalog = true;
            continue;
        }
        if in_catalog && line.starts_with("## ") && !line.contains("Catálogo") && !line.contains("Catalogo") {
            break;
        }
        if !in_catalog || !line.trim().starts_with('|') {
            continue;
        }
        if Regex::new(r"^\|\s*[-:]+").unwrap().is_match(line) {
            continue;
        }
        if line.contains("Archivo fuente") || line.trim() == "| name |" {
            continue;
        }
        let inner = line.trim().trim_start_matches('|').trim_end_matches('|');
        let cells: Vec<_> = inner.split('|').map(str::trim).filter(|s| !s.is_empty()).collect();
        if cells.is_empty() {
            continue;
        }
        let first = cells[0];
        let token = if let Some(caps) = re_backtick.captures(first) {
            caps[1].to_string()
        } else {
            first.to_string()
        };
        if token.is_empty() || token.eq_ignore_ascii_case("archivo fuente") || token == "name" {
            continue;
        }
        let name = if token.ends_with(".md") {
            token
        } else {
            format!("{token}.md")
        };
        indexed.insert(name);
    }
    indexed
}

fn audit_legacy_tools_dirs(repo: &Path) -> Vec<String> {
    let mut errors = Vec::new();
    fn walk(dir: &Path, repo: &Path, errors: &mut Vec<String>) {
        let Ok(rd) = fs::read_dir(dir) else { return };
        for entry in rd.flatten() {
            let path = entry.path();
            if path.is_dir() {
                if path.file_name().and_then(|s| s.to_str()) == Some(".SddIA") {
                    if let Ok(names) = fs::read_dir(&path) {
                        for child in names.flatten() {
                            if child.file_name().to_string_lossy() == LEGACY_LOCAL_DIRNAME {
                                let legacy = child.path();
                                errors.push(format!(
                                    "Violación de Simetría Fractal: {} — exigido `{}` en minúscula estricta",
                                    legacy.display(),
                                    path.join("tools").display()
                                ));
                            }
                        }
                    }
                }
                walk(&path, repo, errors);
            }
        }
    }
    walk(repo, repo, &mut errors);
    errors
}

fn audit_scope(label: &str, tools_dir: &Path, index_path: &Path) -> Vec<String> {
    let mut errors = Vec::new();
    if !tools_dir.is_dir() && !index_path.is_file() {
        return errors;
    }
    if index_path.is_file() && !tools_dir.is_dir() {
        errors.push(format!("{label}: index exists but directory missing: {}", tools_dir.display()));
        return errors;
    }
    if tools_dir.is_dir() && !index_path.is_file() {
        errors.push(format!("{label}: definitions present but index missing: {}", index_path.display()));
        return errors;
    }
    let on_disk = scan_definitions(tools_dir);
    let in_index = parse_index_filenames(index_path);
    for name in on_disk.difference(&in_index) {
        errors.push(format!("{label}: orphan file (not in index): {}/{}", tools_dir.display(), name));
    }
    for name in in_index.difference(&on_disk) {
        errors.push(format!("{label}: orphan index row (no file): {name}"));
    }
    errors
}

pub fn run(repo: &Path) -> i32 {
    let mut errors = Vec::new();
    errors.extend(audit_legacy_tools_dirs(repo));
    let core_dir = repo.join("SddIA/tools");
    errors.extend(audit_scope("core", &core_dir, &core_dir.join("index.md")));
    let local_dir = repo.join(".SddIA/tools");
    let local_index = repo.join(".SddIA/tools/index.md");
    if local_index.is_file() || local_dir.is_dir() {
        errors.extend(audit_scope("local", &local_dir, &local_index));
    }
    if errors.is_empty() {
        println!("verify-tools-index: OK");
        0
    } else {
        eprintln!("verify-tools-index: FAILED");
        for e in errors {
            eprintln!("{e}");
        }
        1
    }
}
